crates/ruma-state-res/src/lib.rs

```diff
@@ -194,17 +194,17 @@ pub fn separate(
 }
 
 /// Returns a Vec of deduped EventIds that appear in some chains but not others.
-pub fn get_auth_chain_diff(auth_chain_sets: Vec<HashSet<EventId>>) -> HashSet<EventId> {
-    if let Some(first) = auth_chain_sets.first().cloned() {
-        let common = auth_chain_sets
-            .iter()
-            .skip(1)
-            .fold(first, |a, b| a.intersection(b).cloned().collect::<HashSet<EventId>>());
-
-        auth_chain_sets.into_iter().flatten().filter(|id| !common.contains(id)).collect()
-    } else {
-        HashSet::new()
+pub fn get_auth_chain_diff(
+    auth_chain_sets: Vec<HashSet<EventId>>,
+) -> impl Iterator<Item = EventId> {
+    let num_sets = auth_chain_sets.len();
+
+    let mut id_counts: HashMap<EventId, usize> = HashMap::new();
+    for id in auth_chain_sets.into_iter().flatten() {
+        *id_counts.entry(id).or_default() += 1;
     }
+
+    id_counts.into_iter().filter_map(move |(id, count)| (count < num_sets).then(move || id))
```

Comment:
`flatten` is ok because it still iterates over each duplicate and if the `EventId` isn't in one of the sets the count is different?

Should / could the comparison be `count != num_sets`?

Reply:
> `flatten` is ok because it still iterates over each duplicate and if the `EventId` isn't in one of the sets the count is different?

Sounds right.

Re. `!=` vs `<`, it can't possibly be higher, so it's really the same thing 🤷🏼

Reply:
> Re. != vs <, it can't possibly be higher, so it's really the same thing 🤷🏼

Good point haha.